### ISSUE TITLE
major package management fuckage, may have to delete tags

### ISSUE THREAD

**Issue by ExpandingMan** [ExpandingMan/Arrow.jl#32]
@davidanthoff @kristofferC

Hey guys, sorry about this but I seem to have done a number of things to this package which seem to have conspired to completely screw up this package, and the only alternative I can see right now would be to remove it from `JuliaRegistries/General` and re-register it.

So, problem the first is that somehow I managed to supply the local `Project.toml` with the wrong UUID.  I'm not quite sure how I did this, as I was careful to use the UUID that seemed to be assigned to it, but I now suspect that when I did this I was using a local configuration that had not used the UUID's generated for `JuliaRegistries/General`.

Problem the second is that *something* got all screwed up by the tagging, I'm not quite sure what.  It probably has something to do with a fact that at some point I deleted a tag, and recreated one with the same name.  When I do `dev Arrow` in 1.0, I wind up with version 0.1.1.  I am then stuck on version 0.1.1 and I cannot `dev Arrow` with anything newer.

So, at this point it seems that I have FUBARed things so badly  that I don't see any choice but to delete all tags and remove this package from the registries, and register it again.  Please let me know if you have any other ideas.

**Comment by ExpandingMan** [comment]
Ok, I think 0.1.1 was tagged before everything got screwed up.  I'm going to try and see if I can resolve this by working from that point, so I'll re-tag now and that shouldn't affect anything.

**Comment by KristofferC** [comment]
Don't do anything rushed. Even if you added a project file with wrong uuid that should not have impacted  how the package is listed in the registry.

You might be hitting https://github.com/JuliaLang/Pkg.jl/issues/645 which is a big in the package manager.

**Comment by ExpandingMan** [comment]
There seems to be all sorts of strangeness, I can't really tell if it's #645.

Primarily two things: first, when I do `dev Arrow` (with no existing Arrow directory) it pulls this commit: `a3f5b5f21900c90a1b5ca4ef61fac51dbbf81279` for no apparent reason (it wasn't even one of the tagged ones).  Last night, when I did `git pull origin master` on Arrow and then again did `dev Arrow` in the package manager, it was giving me unsatisfiable requirements.  Today I'm on a different machine and am currently unable to reproduce that particular result.

What I started to do to attempt a solution is the following:
- Delete `Project.toml` from all tagged commits.  I won't restore it until `PkgDev` gives me that ability.
- Delete the v0.1.2 and v0.2.0 tags (as everything before and including v0.1.1 seems ok).
- Re-tag v0.1.2, currently has an attobot PR [here](https://github.com/JuliaLang/METADATA.jl/pull/16801) with the current commit (compatible with [0.6, \infinity)).
- Push a commit where I change the `REQUIRE` Julia version (to [0.7, infinity)).
- Tag v0.2.0 on that commit (only because I already had a v0.2.0 that I'm trying to replace).

Since none of those tags will have `Project.toml` at all, and all of the tags I've made since I've been having problems will have been replaced I infer that the above procedure will be guaranteed to give me a working package.

(Of course in my stupidity I already made a [v0.2.0 tag](https://github.com/JuliaLang/METADATA.jl/pull/16802) last night having forgotten to bump the Julia version, which mercifully now has the `waiting for author` tag, so I'll cancel that.)

**Comment by KristofferC** [comment]
`dev` has a bug that it might not load the latest commit, you can just git pull the latest one manually though (https://github.com/JuliaLang/Pkg.jl/issues/465).

Never delete tags that have been registered to METADATA. Don't modify the commit for existing versions registered to METADATA. Just push new tags instead.

**Comment by ExpandingMan** [comment]
Since they are already deleted should I proceed with what I was doing?

**Comment by KristofferC** [comment]
Yes it might be that issue. The best thing would be if you could retag the tags with the same commit as registered in METADATA.jl.

**Comment by ExpandingMan** [comment]
I'm really afraid to do that as those have bad `Project.toml` files.  At this point I'm pretty confident that tagging these later commits will work just fine for `METADATA`, but I'm less confident about `JuliaRegistries/General`.

I'll defer to your judgment, but considering all the problems I've had it would make me feel a lot better if we were able to replace those tags with these commits where we are completely confident that there is no package management related strangeness.

**Comment by ExpandingMan** [comment]
Oh, actually come to think of it we really have to try retagging.  In at least one of them I had accidentally uploaded my `Manifest.toml` before I put it in my `.gitignore`.

Do you think this will be ok, or will this absolutely break `JuliaRegistries/General` and I have to figure something else out?

**Comment by KristofferC** [comment]
Having a Manifest in your package is totally ok :)

**Comment by ExpandingMan** [comment]
By the way, why are the git commit hashes that appear in [METADATA](https://github.com/JuliaLang/METADATA.jl/blob/metadata-v2/Arrow/versions/0.1.0/sha1) completely different from those in [JuliaRegistries/General](https://github.com/JuliaRegistries/General/blob/master/A/Arrow/Versions.toml)?  The latter don't even seem to be commit hashes at all, care to illuminate what's going on there?

**Comment by KristofferC** [comment]
They are tree hashes, i.e. based on the content in the repository instead of the history of it.

**Comment by ExpandingMan** [comment]
Oh wow.  I noticed that you talked about that extensively in your JuliaCon talk, but I didn't realize that it meant that you can't use commit hashes.

**Comment by ExpandingMan** [comment]
Ok, I restored the original tags.  Initially I was having the same problem, but I then deleted my entire `.julia` directory and now things seem restored to normal.

If I may make a suggestion, while I have no doubt that whatever happened here was primarily my fault (perhaps with the flames of my confusion stoked by some `Pkg3` bugs), it does seem that it may be relatively easy for those of us uncomfortable and unfamiliar with the immutability of tagging our own packages (that's right, I seldom do it) to arrive in this sorry state; so it would be really nice if some sort of `unfuck PackageName` command were available.  What I imagine this doing is
- Syncing all registries.
- Completely deleting all references in `clones`, `packages` and `environments` to the package.
- Resolving all packages that depend on that package.
If I'm understanding things correctly (which I may well not be) having such a command would preclude ever having to delete a depot.
